pkg/secret/secret.go

```diff
@@ -0,0 +1,165 @@
+	GetUnused(namespace string, secrets []v1.Secret) (unusedSecrets []v1.Secret, funcErr error)
+	Delete(secrets []v1.Secret)
+	FilterByTime(secrets []v1.Secret, olderThan time.Time) (filteredSecrets []v1.Secret)
+	FilterByMaxCount(secrets []v1.Secret, keep int) (filteredSecrets []v1.Secret)
+	Print(secrets []v1.Secret, batch bool)
+}
+
+//Constructor
+func NewSecretsService(client core.SecretInterface, helper kubernetes.Kubernetes) Service {
+	return SecretsService{
+		client: client,
+		helper: helper,
+	}
+}
+
+// ListNamesAndLabels return the names and labels of all secrets
+func (ss SecretsService) ListNamesAndLabels() (resourceNames, labels []string, err error) {
+	secrets, err := ss.List(metav1.ListOptions{})
+	var objectMetas []metav1.ObjectMeta
+	for _, s := range secrets {
+		objectMetas = append(objectMetas, s.ObjectMeta)
+	}
+	secretNames, labels := util.GetNamesAndLabels(objectMetas)
+	return secretNames, labels, nil
+}
+
+// List returns a list of secrets from a namespace
+func (ss SecretsService) List(listOptions metav1.ListOptions) ([]v1.Secret, error) {
+	secrets, err := ss.client.List(listOptions)
+	if err != nil {
+		return nil, err
+	}
+	return secrets.Items, nil
+}
+
+// GetUnused returns unused resources
+func (ss SecretsService) GetUnused(namespace string, resources []v1.Secret) (unusedResources []v1.Secret, funcErr error) {
+	var usedSecrets []v1.Secret
+	funk.ForEach(openshift.PredefinedResources, func(predefinedResource schema.GroupVersionResource) {
+		funk.ForEach(resources, func(secret v1.Secret) {
+
+			secretName := secret.GetName()
+
+			if funk.Contains(usedSecrets, secret) {
+				// already marked as existing, skip this
+				return
+			}
+			contains, err := ss.helper.ResourceContains(namespace, secretName, predefinedResource)
+			if err != nil {
+				funcErr = err
+				return
+			}
+
+			if contains {
+				usedSecrets = append(usedSecrets, secret)
+			}
+		})
+	})
+
+	for _, resource := range resources {
+		if !funk.Contains(usedSecrets, resource) {
+			unusedResources = append(unusedResources, resource)
+		}
+	}
+
+	return unusedResources, funcErr
+}
+
+// Delete removes secrets
+func (ss SecretsService) Delete(secrets []v1.Secret) {
+	for _, resource := range secrets {
+		namespace := resource.GetNamespace()
+		kind := "Secret"
+		name := resource.GetName()
+
+		log.Infof("Deleting %s %s/%s", kind, namespace, name)
+
+		err := ss.client.Delete(name, &metav1.DeleteOptions{})
+
+		if err != nil {
+			log.WithError(err).Errorf("Failed to delete %s %s/%s", kind, namespace, name)
+		}
+	}
+}
+
+//FilterByTime returns secrets which are older than specified date
+func (ss SecretsService) FilterByTime(secrets []v1.Secret, olderThan time.Time) (filteredResources []v1.Secret) {
+	log.WithFields(log.Fields{
+		"olderThan": olderThan,
+		"resources": secrets,
+	}).Debug("Filtering resources older than the specified time")
+
+	for _, resource := range secrets {
+		lastUpdatedDate := resource.GetCreationTimestamp().Time
```

Comment:
This code would fail if there's no CreationTimestamp set. Please make the code robuster by checking that first, and make a comment about the behaviour of filtering with missing timestamp. I'd suggest to read what GetCreationTimestamp documents, and if not specified, treat as if resources were created 1st of January 1970 (start of Unix timestamps), thus be returned in the result set.